# UCL AI Society Machine Learning Tutorials
### Session 01. Introduction to Numpy, Pandas and Matplotlib Libraries

### Contents
1. Numpy
2. Pandas
3. Matplotlib
4. EDA(Exploratory Data Analysis)

### Aim
At the end of this session, you will be able to:
- Understand the basics of numpy.
- Understand the basics of pandas.
- Understand the basics of matplotlib.
- Build a simple EDA using above libraries.

## 2. Pandas
Pandas is another essential open-source library of Python and today widely used by data scientists and analysts. Wes McKinney built it on the numpy to transform and analyze data. According to the Wikipedia, It is derived from the term "Panel Data", an econometrics term for data sets that include observations over multiple time periods for the same individuals.

### 2.1 Basics of Pandas

In [1]:
# run this shell if you haven't install pandas library
! pip install pandas

In [2]:
import pandas as pd
import numpy as np

In [3]:
print(pd.__version__)

0.24.2


The main data structures of pandas are **Series** and **DataFrames**, where data are stored and manipulated. A `Series` is a column and a `DataFrame` is a multi-dimensional table consisted of a collection of Series.

In [4]:
a = pd.Series([1, 2, 3, np.nan, 5, 6])

In [5]:
print(a)

0    1.0
1    2.0
2    3.0
3    NaN
4    5.0
5    6.0
dtype: float64


In [6]:
b = {
    'Name' : ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune"],
    'Satellite' : [0, 0, 1, 2, 79, 60, 27, 14],
    'AU' : [0.4, 0.7, 1, 1.5, 5.2, 9.5, 19.2, 30.1],
    'Diameter (in 1Kkm)' : [4.9, 12.1, 12.7, 6.8, 139.8, 116.5, 50.7, 49.2]
}

In [7]:
solar_system = pd.DataFrame(b, index = [i for i in range(1, 9)])
solar_system

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1
3    Earth          1   1.0                12.7
4     Mars          2   1.5                 6.8
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [8]:
solar_system.dtypes # check data type

Name                   object
Satellite               int64
AU                    float64
Diameter (in 1Kkm)    float64
dtype: object

We can select what to read from the DataFrame
- `head()` : Extracts the first few data
- `tail()` : Extracts the last few data
- `index` : Extracts the index
- `columns` : Extracts the column
- `loc` : Extracts the information of that row
- `values` : Extracts only the values
- `describe()` : Outputs the summary of statistics of DataFrame
- `sort_values(self, by, axis = 0, ascending = True, inplace = False)` : Sort the DataFrame
- `drop()` : Drop the selected row

In [9]:
df = solar_system

In [10]:
df.head() # the default value in the bracket is 5

Name  Satellite   AU  Diameter (in 1Kkm)
1  Mercury          0  0.4                 4.9
2    Venus          0  0.7                12.1
3    Earth          1  1.0                12.7
4     Mars          2  1.5                 6.8
5  Jupiter         79  5.2               139.8

In [11]:
df.tail(2)

Name  Satellite    AU  Diameter (in 1Kkm)
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [12]:
df.index

Int64Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int64')

In [13]:
df.columns

Index(['Name', 'Satellite', 'AU', 'Diameter (in 1Kkm)'], dtype='object')

In [14]:
df.loc[4]

Name                  Mars
Satellite                2
AU                     1.5
Diameter (in 1Kkm)     6.8
Name: 4, dtype: object

In [15]:
df.values

array([['Mercury', 0, 0.4, 4.9],
       ['Venus', 0, 0.7, 12.1],
       ['Earth', 1, 1.0, 12.7],
       ['Mars', 2, 1.5, 6.8],
       ['Jupiter', 79, 5.2, 139.8],
       ['Saturn', 60, 9.5, 116.5],
       ['Uranus', 27, 19.2, 50.7],
       ['Neptune', 14, 30.1, 49.2]], dtype=object)

In [16]:
df.describe()

Satellite         AU  Diameter (in 1Kkm)
count   8.000000   8.000000             8.00000
mean   22.875000   8.450000            49.08750
std    30.670775  10.853702            52.38417
min     0.000000   0.400000             4.90000
25%     0.750000   0.925000            10.77500
50%     8.000000   3.350000            30.95000
75%    35.250000  11.925000            67.15000
max    79.000000  30.100000           139.80000

In [17]:
df.sort_values(by = 'Diameter (in 1Kkm)', ascending = False)

Name  Satellite    AU  Diameter (in 1Kkm)
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2
3    Earth          1   1.0                12.7
2    Venus          0   0.7                12.1
4     Mars          2   1.5                 6.8
1  Mercury          0   0.4                 4.9

In [18]:
# TO DO: re-sort the DataFrame by the number of satellite in the decreasing order.
None

Before 2016, Pluto was classified as a planet of the solar system. Let's add Pluto to our DataFrame.

In [19]:
df.loc[9] = ["Pluto", 0, 39.5, 2.38]
df

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                4.90
2    Venus          0   0.7               12.10
3    Earth          1   1.0               12.70
4     Mars          2   1.5                6.80
5  Jupiter         79   5.2              139.80
6   Saturn         60   9.5              116.50
7   Uranus         27  19.2               50.70
8  Neptune         14  30.1               49.20
9    Pluto          0  39.5                2.38

Let's excludeit again as it is regraded as a dwarf planet in the field of astronomy.

In [20]:
# Even if the index of Pluto was 9, it is regarded as 8, as the index starts from 0
df.drop(df.index[8])

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1
3    Earth          1   1.0                12.7
4     Mars          2   1.5                 6.8
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

### 2.2 Read Data via Pandas
Pandas supports loading, reading, and writing data from various file format, including CSV, JSON and SQL, by converting it to a DataFrame. 
1. `pd.read_csv()` : Read CSV files
2. `pd.read_json()` : Read JSON files
3. `pd.read_sql_query()` : Read SQL files

In [21]:
# download data at https://support.spatialkey.com/spatialkey-sample-csv-data/
movie = pd.read_csv("./data/IMDB-Movie-Data.csv", index_col = "Title")
movie

Rank                       Genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
Mindhorn                                    8                      Comedy   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Hounds of Love                             23          Crime,Drama,Horror   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Paris pieds nus                            26                      Comedy   
Bahubali: The Beginning                    27      Action,Adventure,Drama   
Dead Awake                                 28             Horror,Thriller   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
...                                       ...                         ...   
Texas Chainsaw 3D                         971             Horror,Thriller   
Disturbia                                 972      Drama,Mystery,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
Dark Places                               977      Drama,Mystery,Thriller   
Amateur Night                             978                      Comedy   
It's Only the End of the World            979                       Drama   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                       985            Adventure,Family   
Your Highness                             986    Adventure,Comedy,Fantasy   
Final Destination 5                       987             Horror,Thriller   
Endless Love                              988               Drama,Romance   
Martyrs                                   989                      Horror   
Selma                                

In [22]:
# To Do: Extract the information of the movie that has the rank of 3
None

In [23]:
# To Do: Re-sort the table in terms of Rating, in the decreasing order
None

In [24]:
# To Do: Re-sort the table in terms of 'Revenue(Millions)', in the increasing order and print out the first 5.
# None
movie.sort_values(by = "Revenue (Millions)", ascending = True).head()

Rank                  Genre  \
Title                                           
A Kind of Murder   232   Crime,Drama,Thriller   
Into the Forest    962  Drama,Sci-Fi,Thriller   
Love, Rosie        678         Comedy,Romance   
Lovesong           322                  Drama   
Wakefield           69                  Drama   

                                                        Description  \
Title                                                                 
A Kind of Murder  In 1960s New York, Walter Stackhouse is a succ...   
Into the Forest   After a massive power outage, two sisters lear...   
Love, Rosie       Rosie and Alex have been best friends since th...   
Lovesong          The relationship between two friends deepens d...   
Wakefield         A man's nervous breakdown causes him to leave ...   

                          Director  \
Title                                
A Kind of Murder      Andy Goddard   
Into the Forest    Patricia Rozema   
Love, Rosie       Christian Ditter   
Lovesong               So Yong Kim   
Wakefield            Robin Swicord   

                                                             Actors  Year  \
Title                                                                       
A Kind of Murder  Patrick Wilson, Jessica Biel, Haley Bennett, V...  2016   
Into the Forest   Ellen Page, Evan Rachel Wood, Max Minghella,Ca...  2015   
Love, Rosie       Lily Collins, Sam Claflin, Christian Cooke, Ja...  2014   
Lovesong          Riley Keough, Jena Malone, Jessie Ok Gray, Car...  2016   
Wakefield         Bryan Cranston, Jennifer Garner, Beverly D'Ang...  2016   

                  Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                    
A Kind of Murder                 95     5.2   3305                0.00   
Into the Forest                 101     5.9  10220                0.01   
Love, Rosie                     102     7.2  80415                0.01   
Lovesong                         84     6.4    616                0.01   
Wakefield                       106     7.5    291                0.01   

                  Metascore  
Title                        
A Kind of Murder       50.0  
Into the Forest        59.0  
Love, Rosie            44.0  
Lovesong               74.0  
Wakefield              61.0

In [25]:
movie['Genre'].value_counts().head(10)

Action,Adventure,Sci-Fi       50
Drama                         48
Comedy,Drama,Romance          35
Comedy                        32
Drama,Romance                 31
Animation,Adventure,Comedy    27
Comedy,Drama                  27
Action,Adventure,Fantasy      27
Comedy,Romance                26
Crime,Drama,Thriller          24
Name: Genre, dtype: int64

In [26]:
movie[movie['Runtime (Minutes)'] <= 120]

Rank                       Genre  \
Title                                                                
Split                                3             Horror,Thriller   
Sing                                 4     Animation,Comedy,Family   
The Great Wall                       6    Action,Adventure,Fantasy   
Mindhorn                             8                      Comedy   
Passengers                          10     Adventure,Drama,Romance   
Moana                               14  Animation,Adventure,Comedy   
Colossal                            15         Action,Comedy,Drama   
The Secret Life of Pets             16  Animation,Adventure,Comedy   
Lion                                19             Biography,Drama   
Arrival                             20        Drama,Mystery,Sci-Fi   
Gold                                21    Adventure,Drama,Thriller   
Hounds of Love                      23          Crime,Drama,Horror   
Trolls                              24  Animation,Adventure,Comedy   
Independence Day: Resurgence        25     Action,Adventure,Sci-Fi   
Paris pieds nus                     26                      Comedy   
Dead Awake                          28             Horror,Thriller   
Bad Moms                            29                      Comedy   
Assassin's Creed                    30      Action,Adventure,Drama   
Why Him?                            31                      Comedy   
Nocturnal Animals                   32              Drama,Thriller   
Deadpool                            34     Action,Adventure,Comedy   
Resident Evil: The Final Chapter    35        Action,Horror,Sci-Fi   
Doctor Strange                      38    Action,Adventure,Fantasy   
5- 25- 77                           40                Comedy,Drama   
Sausage Party                       41  Animation,Adventure,Comedy   
Moonlight                           42                       Drama   
Don't Fuck in the Woods             43                      Horror   
The Founder                         44     Biography,Drama,History   
Lowriders                           45                       Drama   
Fallen                              48     Adventure,Drama,Fantasy   
...                                ...                         ...   
Into the Forest                    962       Drama,Sci-Fi,Thriller   
The Other Boleyn Girl              963     Biography,Drama,History   
I Spit on Your Grave               964       Crime,Horror,Thriller   
Custody                            965                       Drama   
Wrecker                            969      Action,Horror,Thriller   
Texas Chainsaw 3D                  971             Horror,Thriller   
Disturbia                          972      Drama,Mystery,Thriller   
Scream 4                           974              Horror,Mystery   
My Big Fat Greek Wedding 2         976       Comedy,Family,Romance   
Dark Places                        977      Drama,Mystery,Thriller   
Amateur Night                      978                      Comedy   
It's Only the End of the World     979                       Drama   
The Skin I Live In                 980              Drama,Thriller   
Miracles from Heaven               981      Biography,Drama,Family   
Annie                              982         Comedy,Drama,Family   
Let's Be Cops                      984                      Comedy   
Max                                985            Adventure,Family   
Your Highness                      986    Adventure,Comedy,Fantasy   
Final Destination 5                987             Horror,Thriller   
Endless Love                       988               Drama,Romance   
Martyrs                            989                      Horror   
Underworld: Rise of the Lycans     991    Action,Adventure,Fantasy   
Take Me Home Tonight               993        Comedy,Drama,Romance   
Resident Evil: Afterlife           994     Action,Adventure,Horror   
Project X                          995      

In [27]:
# To Do: Extract the movie list whose 'Metascore' is bigger than 70.
None

In [28]:
# To Do: Extract the movie list whose 'Director' is 'Christopher Nolan'
None

#### 2.2.1 Pandas Exercise

To Do: Extract the movie list that meets below requirements:
- 1. Released after (and including) 2010 (key = 'Year')
- 2. Runtime is shorter than 150 minutes (key = 'Runtime (Minutes)')
- 3. Rating is above 8.0 (key = 'Rating') 
Print out only the first 3 movies.

In [29]:
None

### 2.3 Remove Missing Data
To represent missing data, pandas use np.nan. Pandas supports removal of missing data, which may deteriorate the performance of data science or machine learning model. In addition, most of data scientists and machine learning engineers remove missing data when they analyze data. 
- `isnull()` returns True or False, depending on the cell's null status. 
- `sum()` counts the number of nulls in each column.
- `dropna()` deletes any row that contains any single null value.
- `fillna(value)` : Fill missing value with the given values.

In [30]:
movie.isnull()

Rank  Genre  Description  Director  \
Title                                                                          
Guardians of the Galaxy                  False  False        False     False   
Prometheus                               False  False        False     False   
Split                                    False  False        False     False   
Sing                                     False  False        False     False   
Suicide Squad                            False  False        False     False   
The Great Wall                           False  False        False     False   
La La Land                               False  False        False     False   
Mindhorn                                 False  False        False     False   
The Lost City of Z                       False  False        False     False   
Passengers                               False  False        False     False   
Fantastic Beasts and Where to Find Them  False  False        False     False   
Hidden Figures                           False  False        False     False   
Rogue One                                False  False        False     False   
Moana                                    False  False        False     False   
Colossal                                 False  False        False     False   
The Secret Life of Pets                  False  False        False     False   
Hacksaw Ridge                            False  False        False     False   
Jason Bourne                             False  False        False     False   
Lion                                     False  False        False     False   
Arrival                                  False  False        False     False   
Gold                                     False  False        False     False   
Manchester by the Sea                    False  False        False     False   
Hounds of Love                           False  False        False     False   
Trolls                                   False  False        False     False   
Independence Day: Resurgence             False  False        False     False   
Paris pieds nus                          False  False        False     False   
Bahubali: The Beginning                  False  False        False     False   
Dead Awake                               False  False        False     False   
Bad Moms                                 False  False        False     False   
Assassin's Creed                         False  False        False     False   
...                                        ...    ...          ...       ...   
Texas Chainsaw 3D                        False  False        False     False   
Disturbia                                False  False        False     False   
Rock of Ages                             False  False        False     False   
Scream 4                                 False  False        False     False   
Queen of Katwe                           False  False        False     False   
My Big Fat Greek Wedding 2               False  False        False     False   
Dark Places                              False  False        False     False   
Amateur Night                            False  False        False     False   
It's Only the End of the World           False  False        False     False   
The Skin I Live In                       False  False        False     False   
Miracles from Heaven                     False  False        False     False   
Annie                                    False  False        False     False   
Across the Universe                      False  False        False     False   
Let's Be Cops                            False  False        False     False   
Max                                      False  False        False     False   
Your Highness                            False  False        False     False   
Final Destination 5                      False  False        False     False   
Endless Love                             F

In [31]:
movie.isnull().sum()

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

In [32]:
movie.shape

(1000, 11)

In [33]:
movie.fillna(value = 0)

Rank                       Genre  \
Title                                                                       
Guardians of the Galaxy                     1     Action,Adventure,Sci-Fi   
Prometheus                                  2    Adventure,Mystery,Sci-Fi   
Split                                       3             Horror,Thriller   
Sing                                        4     Animation,Comedy,Family   
Suicide Squad                               5    Action,Adventure,Fantasy   
The Great Wall                              6    Action,Adventure,Fantasy   
La La Land                                  7          Comedy,Drama,Music   
Mindhorn                                    8                      Comedy   
The Lost City of Z                          9  Action,Adventure,Biography   
Passengers                                 10     Adventure,Drama,Romance   
Fantastic Beasts and Where to Find Them    11    Adventure,Family,Fantasy   
Hidden Figures                             12     Biography,Drama,History   
Rogue One                                  13     Action,Adventure,Sci-Fi   
Moana                                      14  Animation,Adventure,Comedy   
Colossal                                   15         Action,Comedy,Drama   
The Secret Life of Pets                    16  Animation,Adventure,Comedy   
Hacksaw Ridge                              17     Biography,Drama,History   
Jason Bourne                               18             Action,Thriller   
Lion                                       19             Biography,Drama   
Arrival                                    20        Drama,Mystery,Sci-Fi   
Gold                                       21    Adventure,Drama,Thriller   
Manchester by the Sea                      22                       Drama   
Hounds of Love                             23          Crime,Drama,Horror   
Trolls                                     24  Animation,Adventure,Comedy   
Independence Day: Resurgence               25     Action,Adventure,Sci-Fi   
Paris pieds nus                            26                      Comedy   
Bahubali: The Beginning                    27      Action,Adventure,Drama   
Dead Awake                                 28             Horror,Thriller   
Bad Moms                                   29                      Comedy   
Assassin's Creed                           30      Action,Adventure,Drama   
...                                       ...                         ...   
Texas Chainsaw 3D                         971             Horror,Thriller   
Disturbia                                 972      Drama,Mystery,Thriller   
Rock of Ages                              973        Comedy,Drama,Musical   
Scream 4                                  974              Horror,Mystery   
Queen of Katwe                            975       Biography,Drama,Sport   
My Big Fat Greek Wedding 2                976       Comedy,Family,Romance   
Dark Places                               977      Drama,Mystery,Thriller   
Amateur Night                             978                      Comedy   
It's Only the End of the World            979                       Drama   
The Skin I Live In                        980              Drama,Thriller   
Miracles from Heaven                      981      Biography,Drama,Family   
Annie                                     982         Comedy,Drama,Family   
Across the Universe                       983       Drama,Fantasy,Musical   
Let's Be Cops                             984                      Comedy   
Max                                       985            Adventure,Family   
Your Highness                             986    Adventure,Comedy,Fantasy   
Final Destination 5                       987             Horror,Thriller   
Endless Love                              988               Drama,Romance   
Martyrs                                   989                      Horror   
Selma                                

In [34]:
movie.dropna(inplace = True)

In [35]:
movie.shape

(838, 11)

After removing missing data, the shape of movie dataFrame has changed, from (1000, 11) to (838, 11)

### 2.4 Merging Data
- `concat()` : Concatenation. Used to merge two or more pandas object.
- `merge()` : It is merged in the way of SQL. 

In [36]:
df1 = pd.DataFrame(np.random.randn(10, 2))
df1

0         1
0  0.896173  0.716901
1 -1.997141  0.178740
2 -0.886698  1.481461
3  0.173028 -0.493156
4  0.838700  0.147142
5 -1.277100  0.377638
6  1.498519 -0.453080
7 -1.627298 -0.107200
8  0.149254  0.216347
9  0.468412  0.805645

In [37]:
df2 = pd.DataFrame(np.random.randn(10, 3))
df2

0         1         2
0 -1.382738  0.224039  1.716445
1 -1.009707 -1.754388 -0.271483
2  0.120054  0.266942 -0.316821
3 -1.292154 -0.608201 -0.439765
4  1.514542 -0.550726  1.660942
5 -1.092151 -1.566807  0.203487
6  1.210065  0.305923 -0.318262
7 -1.072489 -1.402229 -1.447015
8  0.210973 -0.329682 -1.758910
9  0.355878  0.205786 -1.367619

In [38]:
pd.concat([df1, df2])

0         1         2
0  0.896173  0.716901       NaN
1 -1.997141  0.178740       NaN
2 -0.886698  1.481461       NaN
3  0.173028 -0.493156       NaN
4  0.838700  0.147142       NaN
5 -1.277100  0.377638       NaN
6  1.498519 -0.453080       NaN
7 -1.627298 -0.107200       NaN
8  0.149254  0.216347       NaN
9  0.468412  0.805645       NaN
0 -1.382738  0.224039  1.716445
1 -1.009707 -1.754388 -0.271483
2  0.120054  0.266942 -0.316821
3 -1.292154 -0.608201 -0.439765
4  1.514542 -0.550726  1.660942
5 -1.092151 -1.566807  0.203487
6  1.210065  0.305923 -0.318262
7 -1.072489 -1.402229 -1.447015
8  0.210973 -0.329682 -1.758910
9  0.355878  0.205786 -1.367619

In [39]:
pd.concat([df1, df2], axis = 1)

0         1         0         1         2
0  0.896173  0.716901 -1.382738  0.224039  1.716445
1 -1.997141  0.178740 -1.009707 -1.754388 -0.271483
2 -0.886698  1.481461  0.120054  0.266942 -0.316821
3  0.173028 -0.493156 -1.292154 -0.608201 -0.439765
4  0.838700  0.147142  1.514542 -0.550726  1.660942
5 -1.277100  0.377638 -1.092151 -1.566807  0.203487
6  1.498519 -0.453080  1.210065  0.305923 -0.318262
7 -1.627298 -0.107200 -1.072489 -1.402229 -1.447015
8  0.149254  0.216347  0.210973 -0.329682 -1.758910
9  0.468412  0.805645  0.355878  0.205786 -1.367619

In [40]:
demis = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Demis' : [75, 97, 64, 81]}
)
demis

Modules  Demis
0   Bioinformatics     75
1  Robotic Systems     97
2         Security     64
3        Compilers     81

In [41]:
sedol = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Sedol' : [63, 78, 84, 95]})
sedol

Modules  Sedol
0   Bioinformatics     63
1  Robotic Systems     78
2         Security     84
3        Compilers     95

In [42]:
pd.merge(demis, sedol, on = 'Modules')

Modules  Demis  Sedol
0   Bioinformatics     75     63
1  Robotic Systems     97     78
2         Security     64     84
3        Compilers     81     95

In [43]:
#To Do: Define your own dataframe and use functions introducesd above to concatenate them.

### What to do next?
Below websites would be helpful for your further study on pandas library:
- [Pandas official website](https://pandas.pydata.org)
- [10 minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html)
- [Data Wrangling with Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)